examples/custom_rule.jl

```diff
@@ -0,0 +1,266 @@
+@show autodiff(Forward, g, Const(y), Duplicated(x, dx)) # derivative of g w.r.t. x[1], with y annotated Const
+@show autodiff(Forward, g, Const(y), Const(x)); # derivative of g w.r.t. x[1], with x and y annotated Const
+
+# Note that there are also exist batched duplicated annotations for forward mode, namely [`BatchDuplicated`](@ref)
+# and [`BatchDuplicatedNoNeed`](@ref), which are not covered in this tutorial.
+
+# ## Defining a reverse-mode rule
+
+# Let's look at how to write a simple reverse-mode rule! 
+# First, we write a method for [`EnzymeRules.augmented_primal`](@ref):
+
+function augmented_primal(config::ConfigWidth{1}, func::Const{typeof(f)}, ::Type{<:Active},
+                          y::Duplicated, x::Duplicated)
+    println("In custom augmented primal rule.")
+    ## Compute primal
+    if needs_primal(config)
+        primal = func.val(y.val, x.val)
+    else
+        y.val .= x.val.^2 # y still needs to be mutated even if primal not needed!
+        primal = nothing
+    end
+    ## Save x in tape if x will be overwritten
+    if overwritten(config)[3]
+        tape = copy(x.val) 
+    else
+        tape = nothing
+    end
+    ## Return an AugmentedReturn object with shadow = nothing
+    return AugmentedReturn(primal, nothing, tape)
+end
+
+# Let's unpack our signature for `augmented_primal` :
+# * We accepted a [`EnzymeRules.Config`](@ref) object with a specified width of 1, which means that our rule does not support batched reverse mode.
+# * We annotated `f` with [`Const`](@ref) as usual.
+# * We dispatched on an [`Active`](@ref) annotation for the return value. This is a special annotation for scalar values, such as our return value,
+#   that indicates that that we care about the value's derivative but we need not explicitly allocate a mutable shadow since it is a scalar value.
+# * We annotated `x` and `y` with [`Duplicated`](@ref), similar to our first simple forward rule.
+
+# Now, let's unpack the body of our `augmented_primal` rule:
+# * We checked if the `config` requires the primal. If not, we need not compute the return value, but we make sure to mutate `y` in all cases.
+# * We checked if `x` could possibly be overwritten using the `Overwritten` attribute of [`EnzymeRules.Config`](@ref). 
+#   If so, we save the value of `x` on the `tape` of the returned [`EnzymeRules.AugmentedReturn`](@ref) objet.
+# * We return a shadow of `nothing` since the return value is [`Active`](@ref) and hence does not need a shadow.
+
+# Now, we write a method for [`EnzymeRules.reverse`](@ref):
+
+function reverse(config::ConfigWidth{1}, func::Const{typeof(f)}, dret::Active, tape,
+                 y::Duplicated, x::Duplicated)
+    println("In custom reverse rule.")
+    ## retrieve x value, either from original x or from tape if x may have been overwritten.
+    xval = overwritten(config)[3] ? tape : x.val 
+    ## accumulate into x's shadow, don't assign!
+    x.dval .+= 2 .* xval .* dret.val 
```

Comment:
you should also add the following:

x.dval .+= 2 .* xval .* y.dval
y.dval = 0;

This is because in reverse mode AD you have to sum up the derivatives from all users. Your case accounts for the use inside of sum(...). However if y was read after our function, we need to consider derivatives from that use as well. (accounted for above).